interceptor/intercept.go

```diff
@@ -85,66 +89,179 @@ func (i *Interceptor) Intercept(nextHop string, reqPaymentHash []byte, reqOutgoi
+		// is unknown.
+		if nextHop == nil {
+			nextHop = destination
+		}
 
-				validUntil, err := time.Parse(basetypes.TIME_FORMAT, params.ValidUntil)
+		if nextHop != nil {
+			isConnected, err := i.client.IsConnected(nextHop)
+			if err != nil {
+				log.Printf("IsConnected(%x) error: %v", nextHop, err)
+				return InterceptResult{
+					Action:      INTERCEPT_FAIL_HTLC_WITH_CODE,
+					FailureCode: FAILURE_TEMPORARY_CHANNEL_FAILURE,
+				}, nil
+			}
+
+			if !isConnected {
+				// If not connected, send a notification to the registered
+				// notification service for this client if available.
+				notified, err := i.notificationService.Notify(
+					hex.EncodeToString(nextHop),
+					reqPaymentHashStr,
+				)
+
+				// If this errors or the client is not notified, the client
+				// is offline or unknown. We'll resume the HTLC (which will
+				// result in UNKOWN_NEXT_PEER)
 				if err != nil {
-					log.Printf("time.Parse(%s, %s) failed. Failing channel open: %v", basetypes.TIME_FORMAT, params.ValidUntil, err)
 					return InterceptResult{
-						Action:      INTERCEPT_FAIL_HTLC_WITH_CODE,
-						FailureCode: FAILURE_TEMPORARY_CHANNEL_FAILURE,
+						Action: INTERCEPT_RESUME,
 					}, nil
 				}
 
-				if time.Now().UTC().After(validUntil) {
-					if !i.isCurrentChainFeeCheaper(token, params) {
-						log.Printf("Intercepted expired payment registration. Failing payment. payment hash: %x, valid until: %s", paymentHash, params.ValidUntil)
+				if notified {
+					log.Printf("Notified %x of pending htlc", nextHop)
+					d, err := time.ParseDuration(i.config.NotificationTimeout)
+					if err != nil {
+						log.Printf("WARN: No NotificationTimeout set. Using default 1m")
+						d = time.Minute
+					}
+					timeout := time.Now().Add(d)
+					err = i.client.WaitOnline(nextHop, timeout)
+
+					// If there's an error waiting, resume the htlc. It will
+					// probably fail with UNKNOWN_NEXT_PEER.
+					if err != nil {
+						log.Printf(
+							"waiting for peer %x to come online failed with %v",
+							nextHop,
+							err,
+						)
 						return InterceptResult{
-							Action:      INTERCEPT_FAIL_HTLC_WITH_CODE,
-							FailureCode: FAILURE_TEMPORARY_CHANNEL_FAILURE,
+							Action: INTERCEPT_RESUME,
 						}, nil
 					}
 
-					log.Printf("Intercepted expired payment registration. Opening channel anyway, because it's cheaper at the current rate. paymenthash: %s, params: %+v", reqPaymentHashStr, params)
+					log.Printf("Peer %x is back online. Continue htlc.", nextHop)
+					// At this point we know a few things.
+					// - This is either a channel partner or a registered payment
+					// - they were offline
+					// - They got notified about the htlc
+					// - They came back online
+					// So if this payment was not registered, this is a channel
+					// partner and we have to wait for the channel to become active
+					// before we can forward.
+					if !isRegistered {
```

Comment:
In general I think we can see that if we don't have a registration in place we can wait for channel to be active from the beginning and only if we have registration we will wait for the peer to come online.
This could be also for a mobile user, not necessary a partner.

Reply:
- with channel partner I mean anyone we have a channel with (could be a mobile user. It's actually probably a mobile user).
- I do think we want to wait for 'normal forwards' right? Not just forwards where we need to open a channel.
- In general I think we only want to wait for the channel to become active if we've sent a notification. In any other case we want to fail fast.